File: app/src/main/java/com/custom/app/ui/login/LoginActivity.java
package com.custom.app.ui.login;

import android.accounts.Account;
import android.accounts.AccountAuthenticatorResponse;
import android.accounts.AccountManager;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;

import androidx.constraintlayout.widget.ConstraintLayout;

import com.base.app.ui.base.BaseActivity;
import com.core.app.BuildConfig;
import com.core.app.util.ActivityUtil;
import com.core.app.util.AlertUtil;
import com.core.app.util.Util;
import com.custom.app.R;
import com.custom.app.data.model.login.LoginResponse;
import com.custom.app.ui.home.HomeActivity;
import com.custom.app.ui.password.forgot.ForgotPasswordDialog;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.gson.Gson;
import com.mobsandgeeks.saripaar.ValidationError;
import com.mobsandgeeks.saripaar.Validator;
import com.mobsandgeeks.saripaar.Validator.ValidationListener;
import com.mobsandgeeks.saripaar.annotation.NotEmpty;
import com.mobsandgeeks.saripaar.annotation.Order;
import com.user.app.data.UserData;

import org.threeten.bp.Instant;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnLongClick;
import dagger.android.AndroidInjection;
import timber.log.Timber;

import static android.accounts.AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE;
import static android.accounts.AccountManager.KEY_ACCOUNT_NAME;
import static android.accounts.AccountManager.KEY_ACCOUNT_TYPE;
import static com.custom.app.util.Constants.FORGOT_PWD_DIALOG;
import static com.network.app.oauth.AccountConstants.KEY_REFRESH_TOKEN;
import static com.network.app.oauth.AccountConstants.KEY_TOKEN_EXPIRE_IN;
import static com.network.app.oauth.AccountConstants.KEY_TOKEN_TYPE;
import static com.network.app.oauth.AccountConstants.TOKEN_EXPIRE_IN_MILLIS;
import static com.network.app.oauth.AccountConstants.TYPE_FULL_ACCESS;

public class LoginActivity extends BaseActivity implements LoginView, ValidationListener {

    private Validator validator;
    private Bundle resultBundle;
    private String authTokenType;
    private AccountAuthenticatorResponse authenticatorResponse;

    @Inject
    LoginPresenter presenter;

    @Inject
    AccountManager accountManager;

    @BindView(R.id.layout_main)
    ConstraintLayout layoutMain;

    @Order(1)
    @NotEmpty(sequence = 1, trim = true, messageResId = R.string.empty_username_msg)
//  @Username(sequence = 2, messageResId = R.string.invalid_username_msg)
    @BindView(R.id.et_username)
    EditText etUsername;

    @Order(2)
    @NotEmpty(sequence = 1, trim = true, messageResId = R.string.empty_password_msg)
//  @Password(sequence = 2, min = 8, scheme = ALPHA_NUMERIC, messageResId = R.string.invalid_password_msg)
    @BindView(R.id.et_password)
    EditText etPassword;

    @BindView(R.id.btn_login)
    Button btnLogin;

    @BindView(R.id.progressBar)
    ProgressBar progressBar;

    @OnClick(R.id.btn_login)
    void login(View view) {
        Util.hideSoftKeyboard(view);
        validator.validate();
    }

    @OnClick(R.id.tv_forgot_pwd)
    void forgotPassword() {
        ForgotPasswordDialog.newInstance().show(getSupportFragmentManager(), FORGOT_PWD_DIALOG);
    }

    @OnClick(R.id.tv_signup)
    void contactAdmin() {
        Util.handleUrl(this, "mailto:admin@custom.com");
    }

    @BindView(R.id.fab_ghost)
    FloatingActionButton fabGhost;

    @OnLongClick(R.id.title)
    boolean toggleFabGhost() {
        if (BuildConfig.DEBUG) {
            if (fabGhost.getVisibility() == View.VISIBLE) {
                fabGhost.hide();
            } else {
                fabGhost.show();
            }
        }
        return true;
    }

    @OnClick(R.id.fab_ghost)
    public void showGhostMenu(View view) {
        super.showGhostMenu(view);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        ButterKnife.bind(this);

        authenticatorResponse = getIntent().getParcelableExtra(KEY_ACCOUNT_AUTHENTICATOR_RESPONSE);

        if (authenticatorResponse != null) {
            authenticatorResponse.onRequestContinued();
        }

        authTokenType = getIntent().getStringExtra(KEY_TOKEN_TYPE);

        if (authTokenType == null) {
            authTokenType = TYPE_FULL_ACCESS;
        }

        validator = new Validator(this);
        validator.setValidationMode(Validator.Mode.IMMEDIATE);
//      validator.registerAnnotation(Username.class);
        validator.setValidationListener(this);

        presenter.setView(this);
    }

    @Override
    public void onValidationSucceeded() {
        presenter.callLogin(etUsername.getText().toString().trim(), etPassword.getText().toString().trim());
    }

    @Override
    public void onValidationFailed(List<ValidationError> errors) {
        for (ValidationError error : errors) {
            View view = error.getView();
            String[] listMessage = error.getCollatedErrorMessage(this).split("\n");
            String msg = listMessage[0];

            if (view instanceof EditText) {
                ((EditText) view).setError(msg);
                view.requestFocus();
            } else {
                showMessage(msg);
            }
        }
    }

    @Override
    public void showProgressBar() {
        progressBar.setVisibility(View.VISIBLE);
        btnLogin.setVisibility(View.INVISIBLE);
    }

    @Override
    public void hideProgressBar() {
        progressBar.setVisibility(View.INVISIBLE);
        btnLogin.setVisibility(View.VISIBLE);
    }

    @Override
    public void showMessage(String msg) {
        AlertUtil.showSnackBar(layoutMain, msg);
    }

    @Override
    public void showHomeScreen(LoginResponse response) {
        addOrUpdateAccount(response);

        if (!userManager.isLoggedIn()) {
            UserData user = response.getUserData();
            userManager.login(new Gson().toJson(user));
            userManager.setUserId(user.getId());
            userManager.setUsername(user.getUsername());
            userManager.setProfilePath(user.getProfile());

            ActivityUtil.startActivity(this, HomeActivity.class, true);
        } else {
            finish();
        }
    }

    public void addOrUpdateAccount(LoginResponse response) {
        Timber.d("Add or update account...");

        String accountName = response.getUserData().getUsername();

        final Account account = new Account(accountName, getString(R.string.account_type));

        long tokenExpireInMillis = (Instant.now().toEpochMilli()) + TOKEN_EXPIRE_IN_MILLIS;

        Bundle data = new Bundle();
        data.putString(KEY_ACCOUNT_NAME, accountName);
        data.putString(KEY_ACCOUNT_TYPE, getString(R.string.account_type));
        data.putString(KEY_REFRESH_TOKEN, response.getRefreshToken());
        data.putString(KEY_TOKEN_EXPIRE_IN, String.valueOf(tokenExpireInMillis));

        accountManager.addAccountExplicitly(account, null, data);
        accountManager.setAuthToken(account, authTokenType, response.getToken());

        resultBundle = data;
    }

    @Override
    public void finish() {
        if (authenticatorResponse != null) {
            if (resultBundle != null) {
                authenticatorResponse.onResult(resultBundle);
            } else {
                authenticatorResponse.onError(AccountManager.ERROR_CODE_CANCELED, "Canceled");
            }
            authenticatorResponse = null;
        }
        super.finish();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        presenter.destroy();
    }
}

File: app/src/main/java/com/custom/app/ui/login/LoginActivity.kt
package com.custom.app.ui.login

import android.accounts.Account
import android.accounts.AccountAuthenticatorResponse
import android.accounts.AccountManager
import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import androidx.lifecycle.lifecycleScope
import com.base.app.ui.base.BaseActivity
import com.core.app.BuildConfig
import com.core.app.util.ActivityUtil
import com.core.app.util.AlertUtil
import com.core.app.util.Util
import com.custom.app.R
import com.custom.app.databinding.ActivityLoginBinding
import com.custom.app.ui.home.HomeActivity
import com.custom.app.ui.password.forgot.ForgotPasswordDialog.Companion.newInstance
import com.custom.app.util.Constants
import com.google.gson.Gson
import com.network.app.model.login.LoginResponse
import com.network.app.oauth.AccountConstants
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.InternalCoroutinesApi
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import org.threeten.bp.Instant
import timber.log.Timber
import javax.inject.Inject

@AndroidEntryPoint
@InternalCoroutinesApi
class LoginActivity : BaseActivity() {

    @Inject
    lateinit var accountManager: AccountManager

    private var resultBundle: Bundle? = null
    private var authTokenType: String? = null
    private var authenticatorResponse: AccountAuthenticatorResponse? = null

    private val viewModel: LoginViewModel by viewModels()

    private var _binding: ActivityLoginBinding? = null
    private val binding get() = _binding!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)

        authenticatorResponse = intent.getParcelableExtra(AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE)

        if (authenticatorResponse != null) {
            authenticatorResponse!!.onRequestContinued()
        }

        authTokenType = intent.getStringExtra(AccountConstants.KEY_TOKEN_TYPE)

        if (authTokenType == null) {
            authTokenType = AccountConstants.TYPE_FULL_ACCESS
        }

        observeViewModel()

        binding.includeFab.fabGhost.setOnClickListener { view: View? -> showGhostMenu(view) }
        binding.title.setOnLongClickListener { view: View? ->
            showHideGhostFab()
            true
        }

        binding.btnLogin.setOnClickListener { view: View? ->
            Util.hideSoftKeyboard(view)
            login()
        }

        binding.tvForgotPwd.setOnClickListener { view: View? -> newInstance().show(supportFragmentManager, Constants.FORGOT_PWD_DIALOG) }
        binding.tvSignup.setOnClickListener { view: View? -> Util.handleUrl(this, "mailto:admin@custom.com") }
        binding.includeFab.fabGhost.setOnClickListener { view: View? -> showGhostMenu(view) }
    }

    private fun observeViewModel() {
        lifecycleScope.launch {
            viewModel.state.collect {
                when (it) {
                    is LoginState.Loading -> showProgressBar()
                    is LoginState.Success -> {
                        hideProgressBar()
                        showHomeScreen(it.response)
                    }
                    is LoginState.Error -> {
                        hideProgressBar()
                        showMessage(it.msg)
                    }
                }
            }
        }
    }

    private fun login() {
        lifecycleScope.launch {
            val username = binding.etUsername.text.toString().trim { it <= ' ' }
            val password = binding.etPassword.text.toString().trim { it <= ' ' }
            viewModel.requestIntent.send(LoginRequestIntent.Login(username, password))
        }
    }

    override fun showProgressBar() {
        binding.progressBar.visibility = View.VISIBLE
        binding.btnLogin.visibility = View.INVISIBLE
    }

    override fun hideProgressBar() {
        binding.progressBar.visibility = View.INVISIBLE
        binding.btnLogin.visibility = View.VISIBLE
    }

    override fun showMessage(msg: String?) {
        AlertUtil.showSnackBar(binding.layoutMain, msg)
    }

    fun showHomeScreen(response: LoginResponse?) {
        addOrUpdateAccount(response)
        if (!userManager.isLoggedIn) {
            val user = response?.userData
            userManager.login(Gson().toJson(user))
            userManager.userId = user?.id
            userManager.username = user?.username
            userManager.profilePath = user?.profile
            ActivityUtil.startActivity(this, HomeActivity::class.java, true)
        } else {
            finish()
        }
    }

    fun addOrUpdateAccount(response: LoginResponse?) {
        Timber.d("Add or update account...")
        val accountName = response?.userData?.username
        val account = Account(accountName, getString(R.string.account_type))
        val tokenExpireInMillis = Instant.now().toEpochMilli() + AccountConstants.TOKEN_EXPIRE_IN_MILLIS
        val data = Bundle()
        data.putString(AccountManager.KEY_ACCOUNT_NAME, accountName)
        data.putString(AccountManager.KEY_ACCOUNT_TYPE, getString(R.string.account_type))
        data.putString(AccountConstants.KEY_REFRESH_TOKEN, response?.refreshToken)
        data.putString(AccountConstants.KEY_TOKEN_EXPIRE_IN, tokenExpireInMillis.toString())
        accountManager.addAccountExplicitly(account, null, data)
        accountManager.setAuthToken(account, authTokenType, response?.token)
        resultBundle = data
    }

    override fun finish() {
        if (authenticatorResponse != null) {
            if (resultBundle != null) {
                authenticatorResponse!!.onResult(resultBundle)
            } else {
                authenticatorResponse!!.onError(AccountManager.ERROR_CODE_CANCELED, "Canceled")
            }
            authenticatorResponse = null
        }
        super.finish()
    }

    private fun showHideGhostFab(): Boolean {
        if (BuildConfig.DEBUG) {
            if (binding.includeFab.fabGhost.visibility == View.VISIBLE) {
                binding.includeFab.fabGhost.hide()
            } else {
                binding.includeFab.fabGhost.show()
            }
        }
        return true
    }
}